google-cloud-firestore/src/main/java/com/google/cloud/firestore/pipeline/expressions/Function.java

```diff
@@ -0,0 +1,2563 @@
+package com.google.cloud.firestore.pipeline.expressions;
+
+import static com.google.cloud.firestore.pipeline.expressions.FunctionUtils.toExprList;
+
+import com.google.api.core.BetaApi;
+import com.google.api.core.InternalApi;
+import com.google.cloud.firestore.DocumentReference;
+import com.google.common.base.Objects;
+import com.google.common.collect.Lists;
+import com.google.firestore.v1.Value;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * This class defines the base class for Firestore {@link com.google.cloud.firestore.Pipeline}
+ * functions, which can be evaluated within pipeline execution.
+ *
+ * <p>This class also provides a set of static functions that can be used to build expressions for
+ * Firestore pipelines.
+ *
+ * <p>This class offers both function variants that directly operates on string field name and
+ * {@link Expr} instances. Using string variant often leads to more concise code while using {@link
+ * Expr} variants offers more flexibility as {@link Expr} could be result of other functions or
+ * complex expressions.
+ *
+ * <p>You can chain together these static functions to create more complex expressions.
+ */
+@BetaApi
+public class Function implements Expr {
+  private final String name;
+  private final List<Expr> params;
+
+  protected Function(String name, List<? extends Expr> params) {
```

Comment:
I think we should prefer using `ImmutableList` instead of wrapping with `unmodifiableList`. Using `ImmutableList` prevents both upstream changes as well as downstream changes. Whereas, simply wrapping a list, still allows the caller to modify the list. Alternatively, we can always makes copies, but that seems wasteful.

The customer facing API can still use a mutable `List` as parameter, but we should always make a copy internally into a `ImmutableList` internally. Once we have an `ImmutableList`, we no longer need to make copies, and can safely pass them around by reference.

```suggestion
  protected Function(String name, ImmutableList<? extends Expr> params) {
```

Reply:
Done.